internal/config/config_file.go

```diff
@@ -8,23 +8,71 @@ import (
 	"os"
 	"path"
 	"path/filepath"
+	"runtime"
 	"syscall"
 
 	"github.com/mitchellh/go-homedir"
 	"gopkg.in/yaml.v3"
 )
 
 const (
-	GH_CONFIG_DIR = "GH_CONFIG_DIR"
+	GH_CONFIG_DIR   = "GH_CONFIG_DIR"
+	XDG_CONFIG_HOME = "XDG_CONFIG_HOME"
 )
 
+// Config path precedence
+// 1. GH_CONFIG_DIR
+// 2. XDG_CONFIG_HOME
+// 3. AppData (windows only)
+// 4. HOME
 func ConfigDir() string {
-	if v := os.Getenv(GH_CONFIG_DIR); v != "" {
-		return v
+	var path string
+	if a := os.Getenv(GH_CONFIG_DIR); a != "" {
+		path = a
+	} else if b := os.Getenv(XDG_CONFIG_HOME); b != "" {
+		path = filepath.Join(b, "gh")
+	} else if c, err := os.UserConfigDir(); runtime.GOOS == "windows" && err == nil && c != "" {
+		path = filepath.Join(c, "GitHub CLI")
+	} else {
+		d, _ := os.UserHomeDir()
+		path = filepath.Join(d, ".config", "gh")
+	}
+
+	// If the path does not exist try migrating config from default paths
+	if _, err := os.Stat(path); err != nil {
+		autoMigrateConfigDir(path)
 	}
 
-	homeDir, _ := homeDirAutoMigrate()
-	return homeDir
+	return path
+}
+
+// Check default paths (os.UserHomeDir, and homedir.Dir) for existing configs
+// If configs exist then move them to newPath
+// TODO: Remove support for homedir.Dir location in v2
+func autoMigrateConfigDir(newPath string) {
+	var paths []string
+	p1, _ := os.UserHomeDir()
+	paths = append(paths, p1)
+	p2, _ := homedir.Dir()
+	paths = append(paths, p2)
+	for _, path := range paths {
+		if path == "" || path == newPath {
+			continue
+		}
+
+		path := filepath.Join(path, ".config", "gh")
+		if s, err := os.Stat(path); err != nil || !s.IsDir() {
+			continue
+		}
+
+		migrateConfigDir(path, newPath)
+		break
+	}
+}
+
+var migrateConfigDir = func(oldPath, newPath string) {
```

Comment:
Could there be a way to keep this part of the code "live" in tests, i.e. to avoid stubbing it?

Recent example: we used to stub `WriteConfigFile` (a pretty critical piece of functionality) in our tests until recently, but then we had a regression in writing the config file on Windows and our tests didn't catch it because we didn't exercise the filesystem operation. https://github.com/cli/cli/pull/3666

Could we instead have tests that set up/migrate actual config directories on disk, just switch them away from the user's actual home dir location?